e2e/testcases/otel_collector_test.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/taskgroup"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/kinds"
+	ocmetrics "kpt.dev/configsync/pkg/metrics"
+)
+
+const (
+	DefaultMonitorKSA = "default"
+	MonitorGSA        = "e2e-test-metric-writer"
+)
+
+func TestOtelCollectorDeployment(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation1, ntopts.RequireGKE(t))
+	nt.T.Cleanup(func() {
+		nt.MustKubectl("delete", "-f", "../testdata/otel-collector/otel-cm-duplicate-timeseries.yaml", "--ignore-not-found")
+		if t.Failed() {
+			nt.PodLogs("config-management-monitoring", ocmetrics.OtelCollectorName, "", false)
+		}
+	})
+
+	// If Workload Identity enabled on cluster, setup KSA to GSA annotation
+	if workloadPool, err := getWorkloadPool(nt); err != nil {
+		nt.T.Fatal(err)
+	} else if workloadPool != "" {
+		nt.T.Log("Workload identity enabled, adding KSA annotation to use %s service account", MonitorGSA)
+		nt.MustKubectl("annotate", "serviceaccount", "--namespace", ocmetrics.MonitoringNamespace, DefaultMonitorKSA,
+			fmt.Sprintf("iam.gke.io/gcp-service-account=%s@%s.iam.gserviceaccount.com", MonitorGSA, *e2e.GCPProject))
+	}
+
+	nt.T.Log("Restart otel-collector pod to refresh the ConfigMap, log and IAM")
+	nomostest.DeletePodByLabel(nt, "app", "opentelemetry", false)
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := validateOtelCollectorStatusCurrent(nt, false); err != nil {
+		nt.T.Fatal()
+	}
+
+	nt.T.Log("Wait 1 minute for metric exporter to settle")
+	time.Sleep(1 * time.Minute)
+	nt.T.Log("Checking the otel-collector log contains no failure...")
+	err := CheckDeploymentLogHasNoFailure(nt, ocmetrics.OtelCollectorName, ocmetrics.MonitoringNamespace)
```

Comment:
What kind of failure are we expecting not to be here?

Reply:
Is there a success signal that we can validate instead of checking the absence of the log?

Reply:
Here we are expecting no failure at all. A healthy unclogged pipeline should be running with `..everything ready..`, clean of error.

Reply:
The issue with checking for the absence of a failure signal is that we aren't really sure that things are working as expected. For example, if we had waited an extra 10 seconds would we have seen the failure log?

Reply:
🤔 Then we will have to ping the listmetric API then - get the ack for metrics, then check for error free log. That's a different monster, I'll see what I can do.

Reply:
Would be awesome if this test performed a GCM metrics query and validated the result.

Reply:
Added list request to verify that metrics have reached GCM. Right now I'm only using the apply_operations_total and checking the cluster name & metric name to match.